### DISCUSSION TITLE
Why KeyUP event is triggered three times?

### DISCUSSION THREAD

**Discussion by mvasi90** [fltk-rs/fltk-rs#831] [Q&A] [answered]
Hi,

The `KeyUp` event is triggered three times. How to get it once?
```rust
...
let editor = TextEditor::default();
...
editor.handle(|_, e| {
   match e {
      Event::KeyUp => {
         println!("{:?}", e);
         }
         _ => {}
      }
      false
 });
```

Returns:
```
Push | KeyDown | KeyUp
Push | KeyDown | KeyUp
Push | KeyDown | KeyUp
```

**Update**

Changing it to:

```rust
editor.handle(|_, e| {
   match e {
      Event::KeyUp => {
         println!("{:?}", e);
         true
         }
         _ => {false}
      }
 });
```

Is triggered only once. But this not happens with `KeyDown` event.

**Comment by MoAlyousef** [comment, marked as answer]
Hello

Unhandled events are propagated upwards (to the parent, window, system handler), the window usually propagates it back to whatever widget that might listen to such an event then if unhandled will ignore it. It might  be that KeyDown is only listened to in a handful of text/input based widgets, so you don’t get that echo. While KeyUp is important for widget/menu shortcuts for example. 

P.S. it would be better to use println!("{}", e) than println!("{:?}", e).

**Reply by mvasi90** [reply]
Ah! It's true. It makes sense.

> it would be better to use println!("{}", e) than println!("{:?}", e)

Understood. I thought it was a special struct that doesn't implement `Display` trait.

Thank you.
